action_server_bt/src/action_server_bt.cpp

```diff
@@ -90,48 +94,64 @@ void ActionServerBT::execute(const std::shared_ptr<GoalHandleActionTree> goal_ha
   {
     params_ = param_listener_->get_params();
     register_behavior_trees(params_, factory_, node_);
+    registerNodesIntoFactory(factory_);
   }
 
   // Loop until something happens with ROS or the node completes
   try
   {
-    // No one "own" this blackboard
-    auto global_blackboard = BT::Blackboard::create();
-    // This blackboard will be owned by "MainTree". It parent is global_blackboard
-    auto root_blackboard = BT::Blackboard::create(global_blackboard);
-    auto tree = factory_.createTree(goal->target_tree, root_blackboard);
+    // This blackboard will be owned by "MainTree". It parent is global_blackboard_
+    auto root_blackboard = BT::Blackboard::create(global_blackboard_);
+
+    tree_ = std::make_shared<BT::Tree>();
+    *tree_ = factory_.createTree(goal->target_tree, root_blackboard);
+    current_tree_name_ = goal->target_tree;
 
     // call user defined function after the tree has been created
-    user_cbs_.on_create(tree);
+    onTreeCreated(*tree_);
     groot_publisher_.reset();
-    groot_publisher_ = std::make_shared<BT::Groot2Publisher>(tree, params_.groot2_port);
+    groot_publisher_ = std::make_shared<BT::Groot2Publisher>(*tree_, params_.groot2_port);
 
     // Loop until the tree is done or a cancel is requested
     const auto period = std::chrono::milliseconds(static_cast<int>(1000.0 / params_.behavior_tick_frequency));
     auto loop_deadline = std::chrono::steady_clock::now() + period;
 
+    // operations to be done if the tree execution is aborted, either by
+    // cancel_requested_ or by onLoopAfterTick()
+    auto abort_action = [this, &action_result, &goal_handle](BT::NodeStatus status, const std::string& message) {
+      action_result->node_status = convert_node_status(status);
+      action_result->error_message = message;
+      RCLCPP_WARN(kLogger, action_result->error_message.c_str());
+      tree_->haltTree();
+      goal_handle->abort(action_result);
+      onTreeExecutionCompleted(status, true);
+    };
+
     while (rclcpp::ok() && status == BT::NodeStatus::RUNNING)
     {
       if (cancel_requested_)
       {
-        action_result->error_message = "Action Server canceling and halting Behavior Tree";
-        RCLCPP_ERROR(kLogger, action_result->error_message.c_str());
-        tree.haltTree();
-        goal_handle->canceled(action_result);
+        abort_action(status, "Action Server canceling and halting Behavior Tree");
```

Comment:
Technically this should return `goal_handle->canceled()`. If we move this statement out of the lambda we can support both `canceled` and `aborted`.